Add Parquet Row Group Bloom Filter Support

huaxingao opened the pull request:
Co-Authored-By: Xi Chen [jshmchenxi@163.com](mailto:jshmchenxi@163.com)
Co-Authored-By: Hao Lin [linhao@qiyi.com](mailto:linhao@qiyi.com)
Co-Authored-By: Huaxin Gao [huaxin_gao@apple.com](mailto:huaxin_gao@apple.com)

Currently, Iceberg has `ParquetMetricsRowGroupFilter` and `ParquetDictionaryRowGroupFilter`. This PR adds one more filter `ParquetBloomRowGroupFilter`, which takes advantage of the Parquet bloom filter to find out if a row group needs to be read or not.

jshmchenxi commented:
@huaxingao Thanks for continuing this work!

hililiwei commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderWithBloomFilter.java:
```diff
@@ -0,0 +1,248 @@
+          .set(TableProperties.PARQUET_VECTORIZATION_ENABLED, "true")
+          .set(TableProperties.PARQUET_BATCH_SIZE, "4")
+          .commit();
+    }
```
I don't quite understand this. Can we inject these properties directly when the table is created instead of changing it after it is created?

```
table = catalog.createTable(TableIdentifier.of("default", name), schema);
```

huaxingao replied:
These properties can be set either at table creation or updated later. I don't think it matters. The reason I picked the second choice is because when I wrote the test, I happened to use `TestSparkReaderDeletes` as my template and followed the style there.

kbendick replied:
+1 to following existing style

hililiwei commented on parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,950 @@
+  public void testNot() {
+    // this test case must use a real predicate, not alwaysTrue(), otherwise binding will simplify it out
+    for (int i = INT_MIN_VALUE - 20; i < INT_MAX_VALUE + 20; i++) {
+      boolean shouldRead = new ParquetBloomRowGroupFilter(SCHEMA,   not(equal("id", i)))
```
format

huaxingao replied:
Fixed. Thanks

hililiwei commented on parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,950 @@
+
+  @Test
+  public void testNot() {
+    // this test case must use a real predicate, not alwaysTrue(), otherwise binding will simplify it out
```
would you mind explaining it?

huaxingao replied:
`TestBloomRowGroupFilter` follows the test style of `TestMetricsRowGroupFilter` and `TestDictionaryRowGroupFilter` and this comment is actually copied. I think what it means is that a real predicate e.g. equal, lessThanOrEqual, etc. need to be used. If you use `alwaysTrue`, it will be simplified out and then the visitEvaluator path won't be tested.

hililiwei commented on api/src/main/java/org/apache/iceberg/data/Record.java:
```diff
@@ -58,4 +58,22 @@ default Record copy(String field1, Object value1, String field2, Object value2,
+                      String field4, Object value4, String field5, Object value5, String field6, Object value6,
+                      String field7, Object value7, String field8, Object value8, String field9, Object value9,
+                      String field10, Object value10, String field11, Object value11) {
+    Map<String, Object> overwriteValues = Maps.newHashMapWithExpectedSize(9);
```
It's more intuitive to just use 12. Although it's actually 12, even with 9.

hililiwei commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -269,19 +282,34 @@ public <D> FileAppender<D> build() throws IOException {
             .withMaxRowCountForPageSizeCheck(rowGroupCheckMaxRecordCount)
-            .build();
+            .withMaxBloomFilterBytes(bloomFilterMaxBytes)
+            .withBloomFilterEnabled(bloomFilterEnabled);
```
When `bloomFilterEnabled` is true, all fields are turned on unless the fields are specified to be false in `columnBloomFilterModes`, right?

huaxingao replied:
That's right

hililiwei reviewed (changes requested):
Great work. Left some comment.

huaxingao commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -167,6 +167,17 @@ private TableProperties() {
+  public static final String PARQUET_BLOOM_FILTER_MAX_BYTES = "write.parquet.bloom-filter-max-bytes";
+  public static final int PARQUET_BLOOM_FILTER_MAX_BYTES_DEFAULT = 1024 * 1024;
+
+
```
Removed. Thanks

huaxingao commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -239,6 +245,13 @@ public <D> FileAppender<D> build() throws IOException {
+          PARQUET_BLOOM_FILTER_MAX_BYTES_DEFAULT);
+      Map<String, String> columnBloomFilterModes = getBloomColumnConfigMap(PARQUET_BLOOM_FILTER_COLUMN_ENABLED_PREFIX);
+      Map<String, String> columnBloomFilterNDVs =
+          getBloomColumnConfigMap(PARQUET_BLOOM_FILTER_COLUMN_EXPECTED_NDV_PREFIX);
```
It's cleaner to get it from the Context. Changed. Thanks

huaxingao commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -292,11 +320,39 @@ public <D> FileAppender<D> build() throws IOException {
+
+        for (Map.Entry<String, String> entry : columnBloomFilterModes.entrySet()) {
+          String col = entry.getKey();
+          String value = entry.getValue();
```
Changed

huaxingao commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -292,11 +320,39 @@ public <D> FileAppender<D> build() throws IOException {
+
+        for (Map.Entry<String, String> entry : columnBloomFilterNDVs.entrySet()) {
+          String col = entry.getKey();
+          String value = entry.getValue();
```
Changed

huaxingao commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -292,11 +320,39 @@ public <D> FileAppender<D> build() throws IOException {
+          .filter(key -> key.startsWith(prefix))
+          .forEach(key -> {
+            String columnAlias = key.replaceFirst(prefix, "");
+            String bloomFilterMode = config.get(key);
```
Changed key to `columnPath`. I left `bloomFilterMode` as is because it could either be `columnBloomFilterEnabled` or `columnBloomFilterNDVs`.

huaxingao commented on parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,950 @@
+  private static final long LONG_BASE = 100L;
+  private static final double DOUBLE_BASE = 1000D;
+  private static final float FLOAT_BASE = 10000F;
+  private static final String BINARY_PREFIX = "BINARY测试_";
```
Actually I prefer to test both English and Chinese (for double bytes characters)

huaxingao commented:
cc @aokolnychyi @RussellSpitzer @flyrain @szehon-ho 
This PR is ready for review. Thank you very much in advance!

huaxingao commented:
also cc @kbendick @chenjunjiedada @rdblue

chenjunjiedada commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -167,6 +167,16 @@ private TableProperties() {
+  public static final boolean DEFAULT_PARQUET_BLOOM_FILTER_ENABLED_DEFAULT = false;
+
+  public static final String PARQUET_BLOOM_FILTER_COLUMN_ENABLED_PREFIX = "write.parquet.bloom-filter-enabled.column.";
+  public static final String PARQUET_BLOOM_FILTER_COLUMN_EXPECTED_NDV_PREFIX =
```
It would be better to document that the NDV is specific for a parquet file.

stevenzwu commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -167,6 +167,16 @@ private TableProperties() {
       "write.delete.parquet.row-group-check-max-record-count";
   public static final int PARQUET_ROW_GROUP_CHECK_MAX_RECORD_COUNT_DEFAULT = 10000;
 
+  public static final String DEFAULT_PARQUET_BLOOM_FILTER_ENABLED = "write.parquet.bloom-filter-enabled.default";
```
does it ever make sense to enable bloom filter for all columns? should we only allow bloom filter for explicitly specified columns?

huaxingao replied:
It is not a common usage to enable bloom filter for all columns, but it's legal. This is consistent with the parquet-mr bloom filter implementations.

rdblue replied:
I agree with Steven. I think it only makes sense to enable bloom filters for some columns.

kbendick replied:
+1. While it's consistent with the parquet-mr bloom filter implementaiton, we need to think of user experience first and foremost.

It doesn't make sense to enable bloom filters for _a lot_ of columns. And many users don't do any tuning of their metadata / statistics.

I think it's in-line with other things we do to make the users experience better, like turning off column level statistics after a certain number of columns. We can point it out in the docs under a big `!!!NOTE` (that's highlighted) that bloom filter is only used when turned on.

It's really an advanced thing to use at all imo.

huaxingao replied:
Sounds good. I have removed `DEFAULT_PARQUET_BLOOM_FILTER_ENABLED`. Now user needs to enable bloom filter for individual column using `PARQUET_BLOOM_FILTER_COLUMN_ENABLED_PREFIX`. If the column is a complex type, user needs to enable the column inside the complex type, for example, 
`set(PARQUET_BLOOM_FILTER_COLUMN_ENABLED_PREFIX + "struct_col.int_field", "true")`

rdblue commented on api/src/main/java/org/apache/iceberg/data/Record.java:
```diff
@@ -58,4 +58,22 @@ default Record copy(String field1, Object value1, String field2, Object value2,
+    overwriteValues.put(field9, value9);
+    overwriteValues.put(field10, value10);
+    overwriteValues.put(field11, value11);
+    return copy(overwriteValues);
```
Do we need to change the public API? And why not use a map for large changes like this?

huaxingao replied:
This change is unnecessary. Removed.

rdblue commented on docs/tables/configuration.md:
```diff
@@ -50,6 +50,10 @@ Iceberg tables support table properties to configure table behavior, like the de
 | write.parquet.compression-codec    | gzip               | Parquet compression codec: zstd, brotli, lz4, gzip, snappy, uncompressed |
 | write.parquet.compression-level    | null               | Parquet compression level                          |
+| write.parquet.bloom-filter-enabled.default | false | Whether to enable writing bloom filter for all columns |
+| write.parquet.bloom-filter-enabled.column.col1 | (not set) | Whether to enable writing bloom filter for column 'col1' to allow per-column configuration; This property overrides `bloom-filter-enabled.default` for the specified column; For example, setting both `write.parquet.bloom-filter-enabled.default=true` and `write.parquet.bloom-filter-enabled.column.some_col=false` will enable bloom filter ...
```
I think this can be a lot shorter. "Enables writing a bloom filter for the column."

huaxingao replied:
Changed. Thanks!

rdblue commented on docs/tables/configuration.md:
```diff
@@ -50,6 +50,10 @@ Iceberg tables support table properties to configure table behavior, like the de
 | write.parquet.compression-level    | null               | Parquet compression level                          |
+| write.parquet.bloom-filter-enabled.default | false | Whether to enable writing bloom filter for all columns |
+| write.parquet.bloom-filter-enabled.column.col1 | (not set) | Whether to enable writing bloom filter for column 'col1' to allow per-column configuration; This property overrides `bloom-filter-enabled.default` for the specified column; For example, setting both `write.parquet.bloom-filter-enabled.default=true` and `write.parquet.bloom-filter-enabled.column.some_col=false` will enable bloom filter ...
+| write.parquet.bloom-filter-expected-ndv.column.col1 | (not set) | The expected number of distinct values in a column, it is used to compute the optimal size of the bloom filter; Note that the NDV is specific for a parquet file. If this property is not set, the bloom filter will use the maximum size set in `bloom-filter-max-bytes`; If this property is set for a column, then no need to enable the ...
```
Shouldn't this come from a table sketch instead of manual configuration? Since we're adding NDV estimation, I think we probably want to avoid setting this.

Also, if this eliminates the need to set `enabled` then we should remove the `enabled` property.

huaxingao replied:
Removed this property.

rdblue commented on docs/tables/configuration.md:
```diff
@@ -50,6 +50,10 @@ Iceberg tables support table properties to configure table behavior, like the de
+| write.parquet.bloom-filter-enabled.default | false | Whether to enable writing bloom filter for all columns |
+| write.parquet.bloom-filter-enabled.column.col1 | (not set) | Whether to enable writing bloom filter for column 'col1' to allow per-column configuration; This property overrides `bloom-filter-enabled.default` for the specified column; For example, setting both `write.parquet.bloom-filter-enabled.default=true` and `write.parquet.bloom-filter-enabled.column.some_col=false` will enable bloom filter ...
+| write.parquet.bloom-filter-expected-ndv.column.col1 | (not set) | The expected number of distinct values in a column, it is used to compute the optimal size of the bloom filter; Note that the NDV is specific for a parquet file. If this property is not set, the bloom filter will use the maximum size set in `bloom-filter-max-bytes`; If this property is set for a column, then no need to enable the ...
+| write.parquet.bloom-filter-max-bytes | 1048576 (1 MB) | The maximum number of bytes for a bloom filter bitset |
```
What is the behavior of this? If the NDV requires a size that is too large, does it skip writing the bloom filter?

huaxingao replied:
If the NDV requires a size that is too large, parquet still writes the bloom filter using the max bytes set by this property, not using the bitset calculated by NDV.

rdblue replied:
I guess there probably isn't much we can do about this, although that behavior makes no sense to me. Is it possible to set the expected false positive probability anywhere? Or is that hard-coded in the Parquet library?

huaxingao replied:
There isn't a property to set fpp in Parquet.

rdblue replied:
What fpp is used by Parquet?

huaxingao replied:
Parquet uses 0.01 for fpp.
I chatted offline with @chenjunjiedada, we can probably add a config for fpp in parquet first, and then in iceberg.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -292,11 +317,39 @@ public <D> FileAppender<D> build() throws IOException {
             .withPageSize(pageSize)
             .withDictionaryPageSize(dictionaryPageSize)
-            .build(),
+            // TODO: add .withMaxBloomFilterBytes(bloomFilterMaxBytes) once ParquetWriter.Builder supports it
```
If this isn't supported, then we should remove the option to begin with, right?

huaxingao replied:
Right. Deleted this comment.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -292,11 +317,39 @@ public <D> FileAppender<D> build() throws IOException {
+          parquetWriteBuilder.withBloomFilterNDV(colPath, Long.valueOf(numDistinctValue));
+        }
+
+        return new ParquetWriteAdapter<>(
```
@aokolnychyi, what do you think about removing the old `ParquetWriteAdapter` code? I don't think that anyone uses it anymore.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -292,11 +317,39 @@ public <D> FileAppender<D> build() throws IOException {
       }
     }
 
+    private static Map<String, String> getBloomColumnConfigMap(String prefix, Map<String, String> config) {
```
Style: method name with `get`.

huaxingao replied:
You mean to remove `get`, right? Changed to `bloomColumnConfigMap`.

rdblue commented on spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/SparkRowLevelOperationsTestBase.java:
```diff
@@ -112,7 +109,17 @@ public static Object[][] parameters() {
+            "parquet",
+            RANDOM.nextBoolean(),
+            WRITE_DISTRIBUTION_MODE_HASH,
+            true
```
Why not add bloom filter testing to one of the existing cases? I don't see a reason why this can't be tested along with other cases. And Spark tests already take a long time. We should avoid adding more cases.

huaxingao replied:
I want to have some bloom filter write/read tests that involve copy on write or merge on read. I think this is the only place that test copy on write or merge on read (I maybe wrong). That's why I added a parameter here to turn on bloom filter to test.

huaxingao replied:
I see. Thanks for the clarification. Instead of executing the entire suite, is it OK to add a new (but small) test suite to test bloom filter for copy on write or merge on read? I didn't find any existing copy on write/merge on read tests with filters.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -106,7 +108,8 @@
           statsFilter.shouldRead(typeWithIds, rowGroup) &&
-              dictFilter.shouldRead(typeWithIds, rowGroup, reader.getDictionaryReader(rowGroup)));
+              dictFilter.shouldRead(typeWithIds, rowGroup, reader.getDictionaryReader(rowGroup)) &&
+                  bloomFilter.shouldRead(typeWithIds, rowGroup, reader.getBloomFilterDataReader(rowGroup)));
```
Nit: this indentation looks suspicious to me. The last line was also after `&&` but this one has a different indentation level.

huaxingao replied:
I changed to  
```
          statsFilter.shouldRead(typeWithIds, rowGroup) &&
              dictFilter.shouldRead(typeWithIds, rowGroup, reader.getDictionaryReader(rowGroup)) &&
              bloomFilter.shouldRead(typeWithIds, rowGroup, reader.getBloomFilterDataReader(rowGroup)));
```
Does this look right now?

rdblue replied:
Yes, looks good.

rdblue commented on parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,950 @@
+
+  @Test
+  public void testIntegerIn() {
+    for (int i = 0; i < INT_VALUE_COUNT; i++) {
```
I don't think this needs to be tested in loops like this. I think the case you want to test is that `in` behaves like `or`. So you should have a case where all values are present, no values are present, and where only one value is present.

huaxingao replied:
Changed. Thanks!

rdblue commented on parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,950 @@
+
+  @Test
+  public void testIntegerNotIn() {
+    for (int i = 0; i < INT_VALUE_COUNT; i++) {
```
Here as well, I think you should test the same cases as `in`.

huaxingao replied:
Changed. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+    public Boolean not(Boolean result) {
+      // not() should be rewritten by RewriteNot
+      // bloom filter is based on hash and cannot eliminate based on not
+      return ROWS_MIGHT_MATCH;
```
Since this calls `rewriteNot` on the filter, you may want to throw an `UnsupportedOperationException`. This is okay as well, but I don't think this will ever be hit.

huaxingao replied:
Fixed. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+        PrimitiveType colType = fileSchema.getType(desc.getPath()).asPrimitiveType();
+        if (colType.getId() != null) {
+          int id = colType.getId().intValue();
+          cols.put(id, desc);
```
It looks like this is only used in `eq` and `in` to check whether the column is in the file or not. But isn't that decision the same as if the column has no bloom filter?

If `bloomCache` doesn't contain the ID, it doesn't matter if it was because the column doesn't exist or if the column exists but has no bloom filter. Either way, this should return `ROWS_MIGHT_MATCH`.

huaxingao replied:
You are right. I deleted `cols`.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+      }
+    }
+
+    private BloomFilter getBloomById(int id) {
```
Style: use of `get` in a method name.

huaxingao replied:
changed to `bloomById`

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+          types.put(id, icebergType);
+        }
+      }
+      if (!hasBloom) {
```
What about when the filter's column set doesn't overlap with any bloom filter columns? I think that would be a stronger check to exit early.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+          Type icebergType = schema.findType(id);
+          boolean noBloom = ParquetUtil.hasNoBloomFilterPages(meta);
+          if (!noBloom) {
+            hasBloom = true;
```
I think `hasBloom` can be removed, in favor of using a `fieldsWithBloomFilter`. Then you can update the check to do a set comparison:

```java
  if (Sets.intersection(fieldsWithBloomFilter, Binder.boundReferences(filter)).isEmpty()) {
    return ROWS_MIGHT_MATCH;
  }
```

huaxingao replied:
Sounds good! Changed the code as you suggested. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+        case DATE:
+        case TIME:
+        case TIMESTAMP:
+          hashValue = bloom.hash(value);
```
Why does this not cast `value`? As long as the type is checked in this method, I think it makes sense to call the right `hash` method, instead of calling a generic one for `Object`.

huaxingao replied:
Casted. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+          return bloom.findHash(hashValue) ? ROWS_MIGHT_MATCH : ROWS_CANNOT_MATCH;
+        case BINARY:
+        case FIXED:
+          hashValue = bloom.hash(Binary.fromReusedByteBuffer((ByteBuffer) value));
```
This is not a reused byte buffer. The contents won't change while `hash` is using it.

huaxingao replied:
Changed to `fromConstantByteBuffer`

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+          switch (col.getPrimitiveType().getPrimitiveTypeName()) {
+            case INT32:
+              hashValue = bloom.hash(decimalValue.unscaledValue().intValue());
+              return bloom.findHash(hashValue) ? ROWS_MIGHT_MATCH : ROWS_CANNOT_MATCH;
```
While I appreciate the use of `ROWS_MIGHT_MATCH` and `ROWS_CANNOT_MATCH`, I think this is no different than returning the result of `findHash` directly, right?

huaxingao replied:
Right. Fixed. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+              int scale = metadata.getScale();
+              int precision = metadata.getPrecision();
+              byte[] requiredBytes = new byte[TypeUtil.decimalRequiredBytes(precision)];
+              byte[] binary = DecimalUtil.toReusedFixLengthBytes(precision, scale, (BigDecimal) value, requiredBytes);
```
The underlying bytes are not reused.

Also, why does this need to copy? If `bloom` works with a Binary you can create it directly from the byte buffer. There should be no need to limit the number of bytes used to `TypeUtil.decimalRequiredBytes`, although if you wanted to be really careful you could check that the expected number of bytes in the buffer matches and return `ROWS_MIGHT_MATCH` if it doesn't.

huaxingao replied:
If I use `decimalValue.unscaledValue().toByteArray()` without padding to the fixed length, I can't find the hash value. 
I took a look at parquet bloom filter's write path, seems it pads the bytes to the fixed length, and then hash.

rdblue replied:
Okay, I see. I think you're right.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+      }
+    }
+
+    private <T> boolean shouldRead(ColumnDescriptor col, T value, BloomFilter bloom, Type type) {
```
Can you move this utility method to the bottom of the class, rather than between override method implementations?

huaxingao replied:
Moved. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,330 @@
+
+    private BloomFilter getBloomById(int id) {
+      if (bloomCache == null) {
+        bloomCache = Maps.newHashMap();
```
Rather than doing this all the time, can you create this cache in `eval`?

huaxingao replied:
Fixed. Thanks!

rdblue commented:
Thanks for working on this, @huaxingao! It looks close overall.

Does it make sense to split this into separate read and write PRs? It seems like adding the evaluator is quite a lot to review independently, and we don't need the write path for its tests.

huaxingao commented:
Thanks @rdblue @kbendick for reviewing!
I have addressed the comments and changed the code accordingly. I changed the code using this PR because I think it might be easier for you to check if I address the comments here. I can split this PR into two afterwards.

rdblue commented on api/src/main/java/org/apache/iceberg/expressions/Binder.java:
```diff
@@ -91,12 +91,21 @@ static Expression bind(StructType struct,
+  }
+
+  public static Set<Integer> boundReferences(
+      StructType struct, List<Expression> exprs, boolean caseSensitive, boolean alreadyBound) {
```
I don't think it is a good idea to add a second boolean argument to this method. It is confusing enough with just one.

How about using a different method name for this and then renaming this to a be a private internal implementation?

huaxingao replied:
I changed this method name to `exprReferences` (please let me know if you have a better name), but I still need to keep this as `public` because I need to access this method from `ParquetBloomRowGroupFilter`, which is in a different package.

rdblue replied:
Maybe just `references` since it doesn't bind?

huaxingao replied:
Sounds good. Will change.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -291,12 +304,32 @@ public <D> FileAppender<D> build() throws IOException {
+      config.keySet().stream()
+          .filter(key -> key.startsWith(prefix))
+          .forEach(key -> {
+            String columnPath = key.replaceFirst(prefix, "");
```
Since this uses column name in the config, is there any logic to update these configs when columns are renamed?

huaxingao replied:
Good question. I don't have a logic to update the configs when columns are renamed. I think we are OK, though. At write path, I use these configs to write bloom filters at file creation time. I don't use these configs any more for read. At read path, the bloom filters are loaded using id instead of column name. If the columns are renamed after the bloom filters have been written, as long as the id are still the same, the bloom filters should be able to loaded OK.

rdblue replied:
We're okay for adding read support, but we should consider how to configure write support then.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,316 @@
+      if (filterRefs.size() > 0 && Sets.intersection(fieldsWithBloomFilter, filterRefs).isEmpty()) {
+        return ROWS_MIGHT_MATCH;
+      }
+      return ExpressionVisitors.visitEvaluator(expr, this);
```
Style: there should be a newline between a control flow block and the following statement.

huaxingao replied:
Changed. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,316 @@
+      if (!fieldsWithBloomFilter.contains(id)) { // no bloom filter
+        return ROWS_MIGHT_MATCH;
+      }
+      BloomFilter bloom = bloomById(id);
```
Style: control flow blocks should be separated from the following block or statement with an empty newline.

huaxingao replied:
Changed. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,316 @@
+      return ROWS_MIGHT_MATCH;
+    }
+
+    private BloomFilter bloomById(int id) {
```
I think `loadBloomFilter` is probably a better name.

huaxingao replied:
Changed to `loadBloomFilter`. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,316 @@
+
+    private <T> boolean shouldRead(PrimitiveType primitiveType, T value, BloomFilter bloom, Type type) {
+      long hashValue = 0;
+      switch (type.typeId()) {
```
I think that this needs to use the Parquet type from the file, not the Iceberg type.

The Iceberg type comes from the read schema, which could be `long` when the file's type is `int` because of type promotion rules. We need to make sure that the hash function used matches the file's type because that was what produced the bloom filter. Hopefully, the same hash function is used for `long` and `int` in Parquet, but I'm not sure that is the case so the safest thing is to convert `value` to the Parquet type.

huaxingao replied:
I changed the code to `switch (primitiveType.getPrimitiveTypeName())`

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,316 @@
+        case DECIMAL:
+          BigDecimal decimalValue = (BigDecimal) value;
+          switch (primitiveType.getPrimitiveTypeName()) {
+            case INT32:
```
I think this is correct.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,316 @@
+              return bloom.findHash(hashValue);
+          }
+        case UUID:
+          hashValue = bloom.hash(Binary.fromReusedByteArray(UUIDUtil.convert((UUID) value)));
```
I don't think this byte array is reused.

huaxingao replied:
Changed to `fromConstantByteArray`. Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,325 @@
+              hashValue = bloom.hash(((Number) value).intValue());
+              return bloom.findHash(hashValue);
+          }
+        case INT64:
```
Looks like this will fall through if the type ID is not handled? I think you need to add a `break;`.

huaxingao replied:
Will fix.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java:
```diff
@@ -0,0 +1,325 @@
+          switch (type.typeId()) {
+            case DECIMAL:
+              BigDecimal decimalValue = (BigDecimal) value;
+              hashValue = bloom.hash(decimalValue.unscaledValue().intValue());
```
This is what Parquet would do to hash a decimal stored as an int?

huaxingao replied:
Yes

rdblue reviewed (commented):
@huaxingao, other than cleaning up `shouldRead` to add avoid falling through for unexpected Iceberg types, I think the read side of this is about ready to go. Do you want to open a separate PR for that?

huaxingao commented:
@rdblue I have a question on how to test the read path. If we separate the write and read path, I don't have a good way to test the read path. Shall we do the write path first? For write path, we can test if the bloom filter is written OK by calling the Parquet bloom filter APIs. Or I just submit the read path PR without a test, and add the test after the write path is in?

rdblue commented:
I'm going to close this because it is broken into separate read and write side PRs.

jia-zhengwei commented:
@huaxingao  
How to create a bloom filter index for iceberg with spark?
I know I need to set (write.parquet.bloom-filter-enabled.column.col1='true') when create table, do I need other operations when insert data by pyspark? 
How to make sure the bloom filter is  OK?  Will they have some new index file generated?
